enclave/core/globals.c

```diff
@@ -8,6 +8,119 @@
+ *        shared-library that has a variable of the same name), hence
+ *        the compiler conservatively uses RIP relative addressing even
+ *        in the current compilation unit.
+ *
+ *     3. extern global:
+ *           extern int g3;
+ *           int get_g3() { return g3; }
+ *
+ *        These are global variables defined in another compilation unit and
+ *        the addresses of these variable can only be known at run time.
+ *        Therefore the compiler always generates an entry in the GOT for
+ *        these variables.
+ *
+ *           get_g3:
+ *              movq g3@GOTPCREL(%rip), %rax
+ *              movl (%rax), %eax
+ *              ret*
+ *
+ * Note, if the source code is compiled using position independent executable
+ * flags (-fPIE or -fpie) instead of position independent code flags
+ * (-fPIC or -fpic), then the variables defined in the compilation unit are
+ * to be given preference over definitions of the same variables in shared
+ * libraries. Hence when compiled using -fPIE/-fpie, local globals
+ * (type 2 above) are also guaranteed to be generated using RIP relative
+ * addressing just like static globals.
+ *
+ *****************************************************************************
+ * Calculating enclave base address securely
+ *
+ * The enclave can be loaded at any address in memory. Only the host knows
+ * the address where the enclave is loaded. So how can an enclave know its
+ * base address:
+ *
+ * Strategy 1:
+ *    Let the host set a variable oe_enclave_base within the enclave after
+ *    loading. This approach is not safe since the host is untrusted and could
+ *    set an arbitrary value and trick the enclave to writing secrets to host
+ *    memory.
+ *
+ * Strategy 2:
+ *    Define a variable called oe_enclave_rva (relative virtual address).
+ *    The oe_enclave_rva is supposed to be set to the offset of that variable
+ *    in the enclave image (i.e. as if the enclave was loaded to memory
+ *    address 0). This offset value is a constant and therefore the enclave
+ *    is signed after setting oe_enclave_rva to this value.
+ *
+ *    Actual runtime address of oe_enclave_rva minus the predetermined offset
+ *    of oe_enclave_rva gives the runtime base address of the enclave:
+ *       enclave-base-address = &oe_enclave_rva - oe_enclave_rva;
+ *
+ *    If the host sets oe_enclave_rva to an arbitrary value, then the loading
+ *    the enclave will fail since the signature (generated with incorrect
+ *    oe_enclave_rva value) will not match the signature embedded by the
+ *    oesign tool (using the correct oe_enclave_rva value).
+ *
+  *****************************************************************************
+ * Applying relocations
+ *
+ * Relocations are done within the enclave and relocation entries are measured
+ * as part of enclave signing. This makes relocations secure.
+ * Relocations require enclave base address to be added to each relocation
```

Comment:
Rephrase: Relocations require that the enclave base address be added to …
Or: Relocations require adding the enclave base address to ...

Reply:
Fixed using style 2. Changed to
Performing relocations involves adding the enclave base address to...